src/routes/calendar.rs

```diff
@@ -1,45 +1,92 @@
-use crate::{middlewares::CookieUserState, scraper::ThinCourse};
+use crate::{
+    middlewares::CookieUserState,
+    scraper::{self, ThinCourse},
+};
+use anyhow::Context;
 use axum::{
     extract::{Json, State},
+    http::{header::SET_COOKIE, HeaderName, StatusCode},
+    response::{AppendHeaders, IntoResponse},
     Extension, Form,
 };
+use axum_extra::extract::cookie::Cookie;
 use maud::{html, Markup};
 use serde::Deserialize;
 use std::sync::Arc;
 use tracing::{debug, instrument};
 
-use super::DatabaseAppState;
+use super::{AppError, DatabaseAppState};
 
 #[derive(Deserialize, Debug)]
 pub struct Search {
-    #[allow(dead_code)] // FIXME: after we actually implement add_to_calendar
-    pub course: String,
+    pub term: String,
+    pub crn: String,
 }
 
-// curl
-// -H "Content-Type: application/x-www-form-urlencoded"
-// -X PUT "http://localhost:8080/calendar"
-// -d "crn=123&crn=456"
-#[instrument(level = "debug", skip(_state))]
-pub async fn add_to_calendar(
-    State(_state): State<Arc<DatabaseAppState>>,
+/// curl
+/// -X PUT "http://localhost:8080/calendar"
+/// -H "Content-Type: application/x-www-form-urlencoded"\
+/// -d "crn=23962&term=202501"
+#[instrument(level = "debug", skip(state))]
+pub async fn add_to_calendar<'a, 'b>(
+    State(state): State<Arc<DatabaseAppState>>,
     Extension(user_state): CookieUserState,
     Form(form): Form<Search>,
-) -> Markup {
-    debug!("add_to_calendar");
-    let mut new_state = user_state.to_owned();
-    new_state.selection.push(ThinCourse {
-        subject_code: "".to_string(),
-        course_code: "".to_string(),
-        sections: Vec::new(),
+) -> Result<impl IntoResponse, AppError> {
+    // get queried term
+    let term: scraper::Term = form.term.parse().map_err(|err| {
+        dbg!(err);
+        StatusCode::BAD_REQUEST
+    })?;
+
+    // get a db conn
+    let conn = state.get_conn(&term).ok_or(StatusCode::NOT_FOUND)?;
+
+    // query db
+    let (subject_code, course_code) = conn
+        .prepare("SELECT subject_code, course_code FROM section WHERE crn = ?1;")
+        .context("failed to prepare courses SQL statement")?
+        .query_row([form.crn], |row| {
+            let sub: String = row.get(0)?;
+            let course: String = row.get(1)?;
+            Ok((sub, course))
+        })
+        .context("query failed")
+        .map_err(|err| {
+            dbg!(err);
```

Comment:
Please use tracing events intead, like `debug!(?err)`. It adds additional context to the console like the current request and function parameters.